Sming/SmingCore/Network/WebsocketClient.cpp

```diff
@@ -0,0 +1,513 @@
+	 |                     Payload Data continued ...                |
+	 +---------------------------------------------------------------+
+
+	 Opcodes
+	 0x00: this frame continues the payload from the last.
+	 0x01: this frame includes utf-8 text data.
+	 0x02: this frame includes binary data.
+	 0x08: this frame terminates the connection.
+	 0x09: this frame is a ping.
+	 0x0A: this frame is a pong.
+	 */
+	uint8_t buffer[500];
+	uint8_t i = 0;
+	buffer[i++] = 0x80 | 0x01; //FINN = 1 and Sending Text Data
+
+	/*
+	 payload_len (7 bits): the length of the payload.
+
+	 0-125 mean the payload is that long.
+	 126 means that the following two bytes indicate the length.
+	 127 means the next 8 bytes indicate the length.
+
+	 So it comes in ~7bit, 16bit and 64bit.
+	 */
+
+	if (length <= 125)
+	{
+		buffer[i++] = 0x80 | length; //Setting MASK bit and length
+	}
+	else
+	{
+		buffer[i++] = 0xFE; // Mask bit + 126
+		buffer[i++] = (uint8_t) (length >> 8);
+		buffer[i++] = (uint8_t) (length & 0xFF);
+	}
+	//64bit outgoing messenges not supported
+
+	byte mask[4];
+	for (uint8_t j = 0; j < 4; j++)
+	{
+		mask[j] = random(0, 255);
+		buffer[i++] = mask[j];
+	}
+	for (uint16_t j = 0; j < length; j++)
+	{
+		buffer[i++] = (msg[j] ^ mask[j % 4]);
+	}
+	send((char*) buffer, i, false);
+}
+
+/* Function Name: sendBinary
+ * Description: Send Binary message to Websocket Server
+ *              Max length of message permitted is 0xffff (65535) bytes
+ * Parameters: msg - Message int8 array 
+ *             length - length of msg  array
+ */
+void WebsocketClient::sendBinary(uint8_t* msg, uint16_t length)
+{
+	uint8_t buffer[500];
+	uint8_t i = 0;
+	buffer[i++] = 0x80 | 0x02; //FINN = 1 and Sending Binary Data
+
+	if (length <= 125)
+	{
+		buffer[i++] = 0x80 | length; //Setting MASK bit and length
+	}
+	else
+	{
+		buffer[i++] = 0xFE; // Mask bit + 126
+		buffer[i++] = (uint8_t) (length >> 8);
+		buffer[i++] = (uint8_t) (length & 0xFF);
+	}
+	//64bit outgoing messenges not supported
+
+	byte mask[4];
+	for (uint8_t j = 0; j < 4; j++)
+	{
+		mask[j] = random(0, 255);
+		buffer[i++] = mask[j];
+	}
+	for (uint16_t j = 0; j < length; j++)
+	{
+		buffer[i++] = (msg[j] ^ mask[j % 4]);
+	}
+	send((char*) buffer, i, false); // Need support from TcpClient to send binary
+}
+
+/* Function Name: onReceive
+ * Description: It process message sent byWebsocket Server
+ *              Max length of message permitted is 0xffff (65535) bytes
+ * Parameters: buf - pbuf array 
+ */
+err_t WebsocketClient::onReceive(pbuf* buf)
+{
+	if (buf == NULL)
+	{
+		// Disconnected, close it
+		TcpClient::onReceive(buf);
+	}
+	else
+	{
+		uint16_t size = buf->tot_len;
+		char* data = new char[size + 1];
+		pbuf_copy_partial(buf, data, size, 0);
+		data[size] = '\0';
+
+		//  debugf("%s", data); //print received buffer
+		switch (Mode)
+		{
+		case ws_Connecting:
+			if (verifyKey(data, size) == true)
+			{
+				Mode = ws_Connected;
+				this->connectedcallback(Mode);
```

Comment:
on my branch connectedcallback() is called after sendPing() I can't remember why